src/district/server/web3_events.cljs

```diff
@@ -6,16 +6,49 @@
             [district.server.web3 :refer [web3]]
             [medley.core :as medley]
             [mount.core :as mount :refer [defstate]]
-            [taoensso.timbre :as log]))
+            [taoensso.timbre :as log]
+            [cljs-node-io.core :as io :refer [slurp spit]]
+            [clojure.edn :as edn]))
 
 (declare start)
 (declare stop)
 
+(def checkpoint-a
+  "Keeps the latest checkpoint (events processed so far).
+  Contains a map with
+  :last-processed-block, a block number
+  :processed-log-indexes, a set of tuples like [tx log-index]"
+  (atom nil))
+
+
+
 (defstate ^{:on-reload :noop} web3-events
   :start (start (merge (:web3-events @config)
                        (:web3-events (mount/args))))
   :stop (stop web3-events))
 
+(defn load-checkpoint-info [file-path]
+  (try
+    (edn/read-string (slurp file-path))
+    (catch js/Error e)))
+
+(defn save-checkpoint! [block-number tx log-index]
+  (swap! checkpoint-a
+         (fn [{:keys [last-processed-block processed-log-indexes] :as checkpoint-info}]
+           (-> checkpoint-info
+               (assoc :last-processed-block block-number
+                      :processed-log-indexes (if (= block-number last-processed-block)
+                                               (conj processed-log-indexes [tx log-index])
+                                               #{[tx log-index]}))))))
+
+
+(defn decorate-dispatch-callback [callback]
```

Comment:
I don't think this is correct naming. decorate is general term, yet the function decorates only with one specific function save-checkpoint. Therefore I think name of the function should be specific as well, such as `add-save-checkpoint`

Reply:
It is called wrap-callback-checkpoint-middleware now.